File: src/main/plugins/configHandler.js
const fs = require('fs');
const path = require('path');
const _ = require('lodash');
const { eleActions } = require('../../shared/constants');
const { getConfigStore, updateConfigStore, isDev, SimpleStore } = require('../functions');

const printStore = new SimpleStore('print_config');

const getLangFilePath = () => {
  if (isDev) {
    return path.join(process.cwd(), 'locales');
  } else {
    return path.join(process.resourcesPath, 'locales');
  }
};

//初始化语言文件
const initLanguageJson = (lang) => {
  const langFilePath = getLangFilePath();
  const langStore = new SimpleStore(lang, langFilePath);
  const defaultLangStore = require(`../locales/${lang}.json`);
  langStore.set(_.merge(defaultLangStore, langStore.get()));
};

//获取所有可用语言
const getAvailableLanguages = () => {
  try {
    const langFilePath = getLangFilePath();
    if (!fs.existsSync(langFilePath)) {
      return [];
    }

    return fs.readdirSync(langFilePath)
      .filter(file => file.endsWith('.json'))
      .map(file => file.replace('.json', ''))
      .filter(lang => lang);
  } catch (e) {
    console.error('Failed to get available languages:', e);
    return [];
  }
};

//读取语言文件
const getLocale = (lang) => {
  try {
    const langFilePath = getLangFilePath();
    const langStore = new SimpleStore(lang, langFilePath);
    return langStore.get();
  } catch (e) {
    console.error(`Failed to read locale ${lang}:`, e);
  }
  return {};
};

const registerConfigAPI = (app, basePath = '/api') => {
  //加载配置
  app.get(`${basePath}/${eleActions.loadConfig}`, (req, res) => {
    try {
      // 初始化语言文件
      initLanguageJson('en');
      initLanguageJson('zh');

      const config = getConfigStore();
      config.Global = config.Global || {};

      // 获取可用语言列表
      config.Global.availableLangs = getAvailableLanguages();

      // 加载所有语言包
      config.Global.locales = {};
      config.Global.availableLangs.forEach(lang => {
        config.Global.locales[lang] = getLocale(lang);
      });
      res.json(config);

    } catch (err) {
      res.status(500).json({
        success: false,
        error: err.message,
      });
    }
  });

  //保存配置
  app.post(`${basePath}/${eleActions.saveConfig}`, (req, res) => {
    try {
      const { Global, Config } = req.body;

      if (!Global && !Config) {
        return res.status(400).json({
          success: false,
          error: 'Global or Config is required',
        });
      }

      if (Global && Config) {
        const configCopy = { ...Config };
        delete configCopy.globalBackground;

        updateConfigStore({
          Config: configCopy,
          Global: _.pick(Global, ['currentLang', 'isShowOverView']),
        });
      }

      console.log('✅ Config saved');
      res.json({ success: true });

    } catch (err) {
      console.error('❌ Save config failed:', err);
      res.status(500).json({
        success: false,
        error: err.message,
      });
    }
  });

  //加载打印配置
  app.get(`${basePath}/${eleActions.loadPrintConfig}`, (req, res) => {
    try {
      const {
        printConfig = {
          scaleX: 100,
          scaleY: 100,
          offsetX: 0,
          offsetY: 0,
        },
      } = printStore.get();
      res.json(printConfig);

    } catch (err) {
      res.status(500).json({
        success: false,
        error: err.message,
      });
    }
  });

  //保存打印配置
  app.post(`${basePath}/${eleActions.savePrintConfig}`, (req, res) => {
    try {
      const { printConfig } = req.body;
      if (!printConfig) {
        return res.status(400).json({
          success: false,
          error: 'printConfig is required',
        });
      }
      printStore.set({printConfig})
      res.json({ success: true });
    } catch (err) {
      console.error('❌ Save print config failed:', err);
      res.status(500).json({
        success: false,
        error: err.message,
      });
    }
  });

};

module.exports = { registerConfigAPI };

File: server/plugins/config_handler.go
package plugins

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"main/server/shared"
	"main/server/storage"
	"main/server/websocket"
	"os"
	"path/filepath"
	"strings"
)

var appVersion string

var printStore *storage.SimpleStore

func init() {
	printStore = storage.NewSimpleStore("print_config")

	// 读取 wails.json
	data, err := os.ReadFile("wails.json")
	if err != nil {
		log.Printf("Failed to read wails.json: %v", err)
		appVersion = "unknown"
		return
	}

	var config struct {
		Info struct {
			ProductVersion string `json:"productVersion"`
		} `json:"info"`
	}

	if err := json.Unmarshal(data, &config); err != nil {
		log.Printf("Failed to parse wails.json: %v", err)
		appVersion = "unknown"
		return
	}
	appVersion = config.Info.ProductVersion
}

// getLangFilePath 获取语言文件路径
func getLangFilePath() string {
	cwd, _ := os.Getwd()
	return filepath.Join(cwd, "locales")
}

// initLanguageJson 初始化语言 JSON
func initLanguageJson(lang string) error {
	cwd, _ := os.Getwd()
	langFilePath := filepath.Join(cwd, "locales")
	langStore := storage.NewSimpleStore(lang, langFilePath)

	defaultLangPath := filepath.Join(cwd, "server", "locales", lang+".json")

	if _, err := os.Stat(defaultLangPath); os.IsNotExist(err) {
		return fmt.Errorf("file not found: %s", defaultLangPath)
	}

	defaultData, err := ioutil.ReadFile(defaultLangPath)
	if err != nil {
		return err
	}

	var defaultLangStore map[string]interface{}
	if err := json.Unmarshal(defaultData, &defaultLangStore); err != nil {
		return err
	}

	currentStore := langStore.Get()
	merged := mergeMaps(defaultLangStore, currentStore)

	if err := langStore.Set(merged); err != nil {
		return err
	}

	return nil
}

// getAvailableLanguages 获取所有可用语言
func getAvailableLanguages() []string {
	langFilePath := getLangFilePath()

	if _, err := os.Stat(langFilePath); os.IsNotExist(err) {
		return []string{}
	}

	files, err := ioutil.ReadDir(langFilePath)
	if err != nil {
		return []string{}
	}

	var langs []string
	for _, file := range files {
		if !file.IsDir() && strings.HasSuffix(file.Name(), ".json") {
			lang := strings.TrimSuffix(file.Name(), ".json")
			if lang != "" {
				langs = append(langs, lang)
			}
		}
	}

	return langs
}

// getLocale 获取语言包
func getLocale(lang string) map[string]interface{} {
	langFilePath := getLangFilePath()
	langStore := storage.NewSimpleStore(lang, langFilePath)
	return langStore.Get()
}

// mergeMaps 合并两个 map
func mergeMaps(base, override map[string]interface{}) map[string]interface{} {
	result := make(map[string]interface{})

	for k, v := range base {
		result[k] = v
	}

	for k, v := range override {
		if vMap, ok := v.(map[string]interface{}); ok {
			if baseMap, ok := result[k].(map[string]interface{}); ok {
				result[k] = mergeMaps(baseMap, vMap)
			} else {
				result[k] = v
			}
		} else {
			result[k] = v
		}
	}

	return result
}

// pickKeys 从 map 中选择指定的 keys
func pickKeys(data map[string]interface{}, keys []string) map[string]interface{} {
	result := make(map[string]interface{})
	for _, key := range keys {
		if val, ok := data[key]; ok {
			result[key] = val
		}
	}
	return result
}

// RegisterConfigHandler 注册配置处理器
func RegisterConfigHandler(wsManager *websocket.Manager) {
	// saveConfig
	wsManager.On(shared.SaveConfig, func(event *websocket.Event, data map[string]interface{}) {
		log.Println("💾 SaveConfig called")

		state, _ := data["state"].(map[string]interface{})
		if state == nil {
			log.Println("⚠️ No state provided")
			return
		}

		Global, _ := state["Global"].(map[string]interface{})
		Config, _ := state["Config"].(map[string]interface{})

		if Global != nil && Config != nil {
			// 删除 globalBackground
			delete(Config, "globalBackground")

			// 只保存特定字段
			pickedGlobal := pickKeys(Global, []string{"currentLang", "isShowOverView"})

			// 保存配置
			err := storage.UpdateConfigStore(map[string]interface{}{
				"Config": Config,
				"Global": pickedGlobal,
			})

			if err != nil {
				log.Printf("❌ Failed to save config: %v\n", err)
			} else {
				log.Println("✅ Config saved")
			}
		}
	})

	// loadConfig
	wsManager.On(shared.LoadConfig, func(event *websocket.Event, data map[string]interface{}) {
		log.Println("📂 LoadConfig called")

		returnChannel, _ := data["returnChannel"].(string)
		log.Printf("   Return channel: %s\n", returnChannel)

		// 初始化语言文件
		if err := initLanguageJson("en"); err != nil {
			log.Printf("⚠️ Failed to init en: %v\n", err)
		}

		if err := initLanguageJson("zh"); err != nil {
			log.Printf("⚠️ Failed to init zh: %v\n", err)
		}

		// 获取配置
		config := storage.GetConfigStore()

		// 确保 Global 存在
		Global, ok := config["Global"].(map[string]interface{})
		if !ok {
			Global = make(map[string]interface{})
			config["Global"] = Global
		}

		// 获取可用语言列表
		Global["availableLangs"] = getAvailableLanguages()

		// 加载所有语言包
		locales := make(map[string]interface{})
		availableLangs, _ := Global["availableLangs"].([]string)
		for _, lang := range availableLangs {
			locales[lang] = getLocale(lang)
		}
		Global["locales"] = locales

		// 发送响应
		log.Printf("📤 Sending config to: %s\n", returnChannel)
		event.Sender.Send(returnChannel, config)
		log.Println("✅ Config sent")
	})

	// savePrintConfig
	wsManager.On(shared.SavePrintConfig, func(event *websocket.Event, data map[string]interface{}) {
		log.Println("🖨️ SavePrintConfig called")

		printConfig, _ := data["printConfig"].(map[string]interface{})
		if printConfig == nil {
			log.Println("⚠️ No printConfig provided")
			return
		}

		err := printStore.Set(map[string]interface{}{
			"printConfig": printConfig,
		})

		if err != nil {
			log.Printf("❌ Failed to save print config: %v\n", err)
		} else {
			log.Println("✅ Print config saved")
		}
	})

	// loadPrintConfig
	wsManager.On(shared.LoadPrintConfig, func(event *websocket.Event, data map[string]interface{}) {
		log.Println("🖨️ LoadPrintConfig called")

		returnChannel, _ := data["returnChannel"].(string)
		log.Printf("   Return channel: %s\n", returnChannel)

		// 获取打印配置
		storeData := printStore.Get()
		printConfig, ok := storeData["printConfig"].(map[string]interface{})

		// 如果不存在，使用默认值
		if !ok {
			printConfig = map[string]interface{}{
				"scaleX":  100,
				"scaleY":  100,
				"offsetX": 0,
				"offsetY": 0,
			}
			log.Println("   Using default print config")
		}

		// 发送响应
		log.Printf("📤 Sending print config to: %s\n", returnChannel)
		event.Sender.Send(returnChannel, printConfig)
		log.Println("✅ Print config sent")
	})

	log.Println("✅ Config handlers registered (3 handlers)")
}
